src/shims/unix/socket.rs

```diff
@@ -20,17 +38,122 @@ impl FileDescription for SocketPair {
         self: Box<Self>,
         _communicate_allowed: bool,
     ) -> InterpResult<'tcx, io::Result<()>> {
+        // This is used to signal socketfd of other side that there is no writer to it's readbuf.
+        // If the upgrade fails, there is no need to update as all read ends has been dropped.
+        if let Some(writebuf) = self.writebuf.upgrade() {
+            writebuf.borrow_mut().buf_has_writer = false;
+        };
         Ok(Ok(()))
     }
+
+    fn read<'tcx>(
+        &mut self,
+        _communicate_allowed: bool,
+        bytes: &mut [u8],
+        ecx: &mut MiriInterpCx<'tcx>,
+    ) -> InterpResult<'tcx, io::Result<usize>> {
+        let request_byte_size = bytes.len();
+        let data_size: usize = self.readbuf.borrow().buf.len();
+        let readbuf_has_writer = self.readbuf.borrow().buf_has_writer;
+
+        // Always succeed on read size 0.
+        if request_byte_size == 0 {
+            return Ok(Ok(0));
+        }
+
+        if self.is_nonblock {
+            if !readbuf_has_writer {
+                if data_size == 0 {
+                    // Non-blocking socketpair with no writer and empty buffer.
+                    return Ok(Ok(0));
+                }
+            } else if data_size == 0 {
+                // Non-blocking socketpair with writer and empty buffer.
+                // https://linux.die.net/man/2/read
+                // EAGAIN or EWOULDBLOCK can be returned for socket,
+                // POSIX.1-2001 allows either error to be returned for this case.
+                // Since there is no ErrorKind for EAGAIN, WouldBlock is used.
+                return Ok(Err(Error::from(ErrorKind::WouldBlock)));
+            }
+        } else {
+            if !readbuf_has_writer {
+                if data_size == 0 {
+                    // Blocking socketpair with no writer and empty buffer.
+                    return Ok(Ok(0));
+                }
+            } else if data_size == 0 {
+                // Blocking socketpair with writer and empty buffer.
+                // FIXME: blocking is currently not supported
+                throw_unsup_format!("socketpair read: blocking isn't supported yet");
+            }
+        }
+        if let Some(data_race) = &ecx.machine.data_race {
+            data_race.acquire_clock(&self.readbuf.clone().borrow_mut().clock, &ecx.machine.threads);
+        }
+        // Do full read / partial read based on the space available.
+        let actual_read_size = request_byte_size.min(data_size);
+        self.readbuf.borrow_mut().buf.read_exact(&mut bytes[..actual_read_size]).unwrap();
+        return Ok(Ok(actual_read_size));
```

Comment:
I wonder if we can delegate the size check to the `VecDeque`

```suggestion
        let actual_read_size = self.readbuf.borrow_mut().buf.read(bytes).unwrap();
        return Ok(Ok(actual_read_size));
```

Reply:
I think this works, but we might need to add ``&mut`` in front of ``bytes``
```rust
let actual_read_size = self.readbuf.borrow_mut().buf.read(&mut bytes).unwrap();
return Ok(Ok(actual_read_size));
```